ulib/LowStar.Literal.fsti

```diff
@@ -0,0 +1,137 @@
+/// the C standard, the user writes ``let x = "🤮"``, then KreMLin will generate
+/// ``const char *x = "\xf0\x9f\xa4\xae"``.
+
+/// String literals as buffers
+/// --------------------------
+
+/// Therefore, in order to talk about the interpretation of a string literal as
+/// a series of bytes, we need to define the serialization of Unicode scalar values
+/// (as returned by ``String.list_of_string``) into bytes. This is a commendable and
+/// noble goal, but instead, we choose to restrict ourselves to the ASCII subset of
+/// UTF-8, where the byte encoding of a scalar value is the identity.
+let is_ascii_char (c: Char.char) = UInt32.lt (Char.u32_of_char c) 0x80ul
+
+let ascii_char = c:Char.char{is_ascii_char c}
+
+let is_ascii_string (s: string) =
+  List.Tot.for_all is_ascii_char (String.list_of_string s)
+
+let ascii_string = s:string{is_ascii_string s}
+
+let for_all_tail #a p (l: list a { Cons? l }): Lemma
+  (requires (List.Tot.for_all p l))
+  (ensures (List.Tot.for_all p (List.Tot.tl l)))
+=
+  ()
+
+let ascii_chars_of_ascii_string (s: ascii_string):
+  l:list ascii_char { List.Tot.length l = String.length s }
+=
+  List.Tot.list_ref (String.list_of_string s)
+
+let u8_of_ascii_char (c: ascii_char): x:UInt8.t{ UInt8.v x = Char.int_of_char c } =
+  let x32 = Char.u32_of_char c in
+  assert_norm (pow2 24 * pow2 8 = pow2 32);
+  Math.Lemmas.modulo_modulo_lemma (UInt32.v x32) (pow2 24) (pow2 8);
+  Int.Cast.Full.uint32_to_uint8 x32
+
+/// This means that if a string literal only contains ASCII, then we can easily
+/// reflect its contents in terms of uint8's, without having to talk about the utf8
+/// encoding.
+/// TODO: lemma: S.index (u8s_of_string s) i = String.index s i
+///   (cannot be proven right now because we don't know much about String.index)
+///   (is this even what we want? should we do everything in terms of list_of_string?)
+let u8s_of_ascii_string (s: ascii_string):
+  ss:Seq.seq UInt8.t { Seq.length ss = List.Tot.length (String.list_of_string s) }
+=
+  let cs = List.Tot.map u8_of_ascii_char (ascii_chars_of_ascii_string s) in
+  Seq.seq_of_list cs
+
+/// Consequently, this function becomes in C a simple cast from ``const char *`` to
+/// ``char *``, since immutable buffers don't (yet) have the ``const`` attribute in
+/// KreMLin. (This is unsavory, and should be fixed later.) This way, a string
+/// literal can be seen as an immutable buffer and passed around as such.
+/// This function checks at extraction-time that its argument is a literal.
+val buffer_of_literal: (s: ascii_string) ->
+  ST.Stack (b: IB.ibuffer UInt8.t)
+    (requires (fun _ ->
+      // TODO: find a way to meta-program the "is literal" check
+      String.length s > 0 /\ // needed because of the precondition on igcmalloc_of_list
+      String.length s < pow2 32))
+    (ensures (fun h0 b h1 ->
+      MB.frameOf b == HS.root /\ // is this really useful?
+      MB.recallable b /\
+      MB.alloc_post_mem_common b h0 h1 (u8s_of_ascii_string s)))
```

Comment:
This spec will lead to a proof of `False` in the client code. For example:

- Use `B.lemma_equal_mem_domains` to derive that the returned buffer is `live` in `h0` (i.e. in the memory before calling this function).
- So now the buffer is both `unused_in` and `live` in `h0`, leading to proof of false via `B.live_not_unused_in`.

Would it be too bad if it was `ST` function, with `frameOf b == HS.root`?

Reply:
thanks for checking the assumption; I don't mind putting this in ST and I think miTLS folks would be ok with it -- after all, if they're using this to replace allocations of immutable buffers, then this would be the expected signature